contracts/extensions/contracts/MilestonesExtension.sol

```diff
@@ -0,0 +1,189 @@
+        uint256 totalAmountPercentage;
+
+        // Loop through the milestones and add them to the milestones array
+        uint256 milestonesLength = _milestones.length;
+        for (uint256 i; i < milestonesLength;) {
+            uint256 amountPercentage = _milestones[i].amountPercentage;
+
+            if (amountPercentage == 0) revert MilestonesExtension_INVALID_MILESTONE();
+
+            totalAmountPercentage += amountPercentage;
+            _milestones[i].status = Status.None;
+            milestones.push(_milestones[i]);
+
+            unchecked {
+                i++;
+            }
+        }
+
+        // Check if the all milestone amount percentage totals to 1e18 (100%)
+        if (totalAmountPercentage != 1e18) revert MilestonesExtension_INVALID_MILESTONE();
+
+        emit MilestonesSet(milestonesLength);
+    }
+
+    /// @notice Submit milestone by the acceptedRecipientId.
+    /// @dev Emits a 'MilestonesSubmitted()' event.
+    /// @param _metadata The proof of work
+    function submitUpcomingMilestone(Metadata calldata _metadata) external virtual {
+        _validateSubmitUpcomingMilestone(msg.sender);
+
+        // Get the milestone and update the metadata and status
+        Milestone storage milestone = milestones[upcomingMilestone];
+        milestone.metadata = _metadata;
+
+        // Set the milestone status to 'Pending' to indicate that the milestone is submitted
+        milestone.status = Status.Pending;
+
+        // Emit event for the milestone
+        emit MilestoneSubmitted(upcomingMilestone);
+    }
+
+    /// @notice Review a pending milestone submitted by the acceptedRecipientId.
+    /// @dev Emits a 'MilestoneStatusChanged()' event.
+    /// @param _milestoneStatus New status of the milestone
+    function reviewMilestone(Status _milestoneStatus) external virtual {
+        _validateReviewMilestone(msg.sender, _milestoneStatus);
+        // Check if the milestone status is pending
+
+        milestones[upcomingMilestone].status = _milestoneStatus;
+
+        emit MilestoneStatusChanged(upcomingMilestone, _milestoneStatus);
+
+        if (_milestoneStatus == Status.Accepted) {
+            upcomingMilestone++;
+        }
+    }
+
+    /// ====================================
+    /// ============ Internal ==============
+    /// ====================================
+
+    function _setProposalBid(address _bidderId, uint256 _proposalBid) internal virtual {
+        if (_proposalBid > maxBid) {
+            // If the proposal bid is greater than the max bid this will revert
+            revert MilestonesExtension_EXCEEDING_MAX_BID();
+        } else if (_proposalBid == 0) {
+            // If the proposal bid is 0, set it to the max bid
+            _proposalBid = maxBid;
+        }
+
+        bids[_bidderId] = _proposalBid;
+        emit SetBid(_bidderId, _proposalBid);
+    }
+
+    function _validateSetMilestones(address _sender) internal virtual {
+        _checkOnlyPoolManager(_sender);
```

Comment:
Why not using `onlyPoolManager` modifier? (check other places too)

Reply:
My rationale was that using the `onlyPoolManager` modifier was too opinionated. I guess there could be use cases in which milestones are not managed by pool managers. I think it's better to modularize the validation part of the code like this in case someone wants to customize it. Wdyt?

Reply:
Good point, then we can make `_checkOnlyPoolManager` (`CoreBaseStrategy.sol`) virtual so they can override and add custom logic. But that is a change for another PR.

Reply:
I think `_checkOnlyPoolManager` is alright as it is. I'd expect devs to modify (in this case) `_validateSetMilestones`. The one to set the milestones might be a MilestonesManager (!= PoolManager).